src/levanter/data/shardcache.py

```diff
@@ -0,0 +1,712 @@
+            if isinstance(restored, Exception):
+                raise restored
+            else:
+                raise RuntimeError().with_traceback(restored[2])
+
+        if global_index < len(self.global_chunks):
+            return self.global_chunks[global_index]
+        elif self._all_done():
+            return None
+        else:
+            # we don't have this chunk yet, so we need to wait
+            if global_index not in self.reader_promises:
+                self.reader_promises[global_index] = asyncio.get_event_loop().create_future()
+            return await self.reader_promises[global_index]
+
+    def _all_done(self):
+        return all(state.finished for state in self.shard_states.values())
+
+    def _check_chunks_ready(self):
+        while all(state.ready() for state in self.shard_states.values()):
+            # get the next chunk for each shard (if it has one)
+            # ordering determined by shard_names
+            old_size = len(self.global_chunks)
+            next_chunks = []
+            # use shard_names for ordering
+            for shard_name in self.shard_source.shard_names:
+                state = self.shard_states[shard_name]
+                if len(state.buffered_chunks) > 0:
+                    next_chunks.append(state.buffered_chunks.pop(0))
+                else:
+                    assert state.finished
+
+            if len(next_chunks) == 0:
+                # all shards are finished, need to notify readers that we are done
+                for k, v in self.reader_promises.items():
+                    assert k >= len(self.global_chunks)
+                    v.set_result(None)
+                self.reader_promises = {}
+
+            # add the next chunk for each shard to the global ordering
+            self.global_chunks += next_chunks
+
+            # Notify any readers that are waiting for this chunk
+            for i in range(old_size, len(self.global_chunks)):
+                if i in self.reader_promises:
+                    self.reader_promises[i].set_result(self.global_chunks[i])
+                    del self.reader_promises[i]
+
+
+def _shard_writer_task(manager: "_ShardCacheManager[T]", shard_names):
+    """This should generally be wrapped in a ray.remote"""
+    try:
+        sources: List[Iterator[T]] = [manager.shard_source.open_shard(shard_name) for shard_name in shard_names]
+        writers = [_ShardWriter(manager.cache_dir, name, manager.config) for name in shard_names]
+        # send any finished chunks to the manager
+        # TODO: probably better to read from a global list of finished chunks
+        i = 0
+        while i < len(writers):
+            writer = writers[i]
+            try:
+                for chunk in writer.written_chunk_metadata:
+                    manager.finished_chunk(writer.shard_name, chunk)
+
+                if writer.is_finished:
+                    manager.finished_shard(writer.shard_name)
```

Comment:
An architectural suggestion would be to do all the source / writer / seek position state management in the actor, and have the remote task just be the write itself i.e. have the calls be unidirectional rather than having the manager and writer call each other. This makes it easier to reason about things like fault tolerance and recovery i.e. we don't have worry about the manager and writer states desynchronizing.

Then you'd have a manager main loop that looks like: wait for any remote write task to complete with [`concurrent.futures.wait(return_when=FIRST_COMPLETED)`](https://docs.python.org/3.3/library/concurrent.futures.html#concurrent.futures.wait), update the source and chunk state and kick off a new remote write task.